### PR TITLE
Use `clap` instead of `getopts` to parse cli args

### PR DESCRIPTION
**Pull request by wert007**
> `clap` allows using subcommands, which are helpfull for the input
> methods. This also makes the main method a lot shorter, since there is
> no need to add every single command line option per manually. This is
> handled by the `derive` feature of `clap` and the new `Args` struct.
>
> commit message

I open this pull request to further discuss things and ensure feature equality, before merging it into main.

Pinging @kevinmatthes

### CODE REVIEW

**Comment by kevinmatthes** [comment]
Would you please post the link to the original repository of `clap` here?  I would like to a look on the implemented traits since I experimented with `getopts` on how the `getopts::Options` and their `getopts::Matches` can be stored in `structs` to thin out the `main`.

The main problem was that `getopts::Options` does not implement neither `Copy` nor `Clone`, so storing the options in a struct is somehow difficult with the `getopts` crate.  But with the `getopts::Matches`, everything is alright.

**Comment by wert007** [comment]
https://github.com/clap-rs/clap

The nice thing here are two things. The first one is, that the cli args are defined in the rust code itself and are turned into the cli at compile time. This is way shorter then declaring every single option with a function call or something similar. The other being, that the input methods are turned into subcommands, which means, that no check is necessary, if it is specified, or if there are multiple specified.

File: src/lib.rs [hunk 1]
```diff
@@ -521,17 +487,58 @@ pub enum LocParseError {
...
-    let name = "commit-analyzer";
-    let synopsis = "[OPTIONS]";
-
-    println!(
-        "{name}.\n{description}\n\n{}",
-        options.usage(&format!("Usage: {name} {synopsis}"))
-    );
+/// Parses the Git history.
+#[derive(Parser, Debug)]
+#[clap(author, version, about, long_about = None)]
```
Comment (kevinmatthes) [inline src/lib.rs:492]: There is no `author` field configured in the `Cargo.toml`.  Please add it or remove the `author` derivation.

File: src/main.rs [hunk 2]
```diff
@@ -1,135 +1,19 @@
...
-        Err(getopts::Fail::UnrecognizedOption(string)) => {
-            eprintln!("{}", opts.usage(&format!("Unknown option '{string}'.")));
-            return sysexits::ExitCode::Usage;
-        }
-        Err(err) => {
-            eprintln!("{}", opts.usage(&format!("{err}")));
-            return sysexits::ExitCode::Config;
-        }
-    };
-    if matches.opt_present("help") {
```
[the remark this answers is not in the document]
Comment (kevinmatthes) [reply src/main.rs:103]: Okay, I found it out myself.  `clap` will show the options, by default.

File: src/lib.rs [hunk 3]
```diff
@@ -521,17 +487,58 @@ pub enum LocParseError {
...
+pub struct Args {
+    #[clap(subcommand)]
+    pub input_method: InputMethod,
```
Comment (kevinmatthes) [inline src/lib.rs:495]: The idea itself is good.  But please use a public getter instead of a public attribute.

File: src/lib.rs (hunk 1 above)
Comment (wert007) [reply src/lib.rs]: Done 👍

File: src/lib.rs (hunk 3 above)
Comment (wert007) [reply src/lib.rs]: Done 👍

**Review by kevinmatthes — CHANGES_REQUESTED** [review]
Thank you for your recent changes!  I left some comments regarding further enhancements.

File: src/lib.rs [hunk 4]
```diff
@@ -336,77 +340,39 @@ impl Filter {
     }
 
     /// Create a new instance from a given set of filter creteria.
-    pub fn new(matches: &getopts::Matches) -> Self {
+    pub fn new(args: Args) -> Self {
```
[the remark this answers is not in the document]
Comment (wert007) [reply src/lib.rs:343]: The disadvantage of borrowing is, that all the vectors, and the strings stored in those vectors need to be cloned, even though, that is actually not necessary. But I try to think of a better solution.

File: src/lib.rs [hunk 5]
```diff
@@ -521,17 +487,86 @@ pub enum LocParseError {
...
+    #[clap(short, long, default_value_t = 3)]
+    duration: u32,
+
+    /// An output file for the commits per day in CSV format.
+    #[clap(short, long)]
+    output: Option<PathBuf>,
+}
+
+impl Args {
+    /// Get a reference to the input method specified by the user.
```
[the remark this answers is not in the document]
Comment (wert007) [reply src/lib.rs:547]: Yeah, I just took the auto generated documentation from rust analyzer and tried to improve it a bit. Will change.

File: src/main.rs [hunk 6]
```diff
@@ -1,135 +1,19 @@
...
-        Some(Err(error)) => {
-            eprintln!("Invalid duration: {error}!");
-            return sysexits::ExitCode::Usage;
-        }
-    };
-    let path = matches.opt_str("output");
-    let commits = match input.read() {
+    let mut args = commit_analyzer::Args::parse();
+
+    let commits = match args.input_method().read() {
```
[the remark this answers is not in the document]
Comment (wert007) [reply src/main.rs:8]: They are subcommands, you can only declare one subcommand at a time, just like in `cargo` or `git`, where `run`, `build`, `commit`, or `rebase` are subcommands. If you try to declare multiple you get an unexpected argument exception.

File: src/lib.rs [hunk 7]
```diff
@@ -521,17 +487,86 @@ pub enum LocParseError {
...
+    #[must_use]
+    pub fn input_method(&self) -> &InputMethod {
+        &self.input_method
+    }
+
+    /// Get the expected verbosity of the program.
+    #[must_use]
+    pub fn is_verbose(&self) -> bool {
+        self.verbose
```
[the remark this answers is not in the document]
Comment (wert007) [reply src/lib.rs:556]: yeah, usually. Thing is, that imho bools should always be written with a verb like `is_`, but a parameter `--is-verbose` would be very ugly. But I check, how one can change the name of the flag.

File: src/lib.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (kevinmatthes) [reply src/lib.rs:343]: How about explicit lifetime annotation and references to the fields in `Args`?  The `Args` instance definitely outlives the creation of the `Filter` when applying the suggestion above.

File: src/lib.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (kevinmatthes) [reply src/lib.rs:547]: Thanks!  By the way another more general thing:  would you please move the struct before the `Author` definition and implementation?  That way, it is easier to take a look at the configured options *and* the structs are sorted by alphabet, again.

File: src/main.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (kevinmatthes) [reply src/main.rs:8]: Okay, thanks!

File: src/lib.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (kevinmatthes) [reply src/lib.rs:556]: I am not such a big fan of these prefixes for variables -- like you use it in the first review comment --, they sound to me more like function / method prefixes; but this is surely a matter of taste.

The easiest thing would be to just rename the method to `verbose` since

```rust
if args.verbose() {
    // ...
}
```

is still of sufficient expressiveness to avoid misunderstandings and fixes the grammar issue arising from the current name.

File: src/lib.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (wert007) [reply src/lib.rs:343]: Hmm, maybe, but this would definitely be a big overhaul then.. I think I have a easier solution, that would be similar to `Args::take_output`, that should be good enough for the time being.

File: src/lib.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (wert007) [reply src/lib.rs]: I decided to just rename the field to `is_verbose` because I personally think it is a way better way to name bool expressions.

File: src/lib.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (wert007) [reply src/lib.rs]: Decided to move not only `Args` to the top, but also `InputMethod` (since this defines the subcommand, and so also the cli) and `Filter` since Filter is also very closely related to the `Args` supplied by the user. So I ordered them by relevance, instead of alphabetical.

File: src/lib.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (wert007) [reply src/lib.rs]: Okay, this was actually a lot easier, since only the vectors needed to be references and not the strings as well. So this has been done with references, also because the other way did not work for some reason.. 😅

File: src/lib.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (kevinmatthes) [reply src/lib.rs]: Please have a look at the following files:

* https://github.com/sorairolake/sysexits-rs/blob/develop/src/lib.rs
* https://github.com/sorairolake/sysexits-rs/blob/develop/src/exit_code.rs

These are the major source files of the `sysexits` crate we use as a dependency.

The maintainer organises their source code by using `lib.rs` for configuration and imports and outsources the other logic to dedicated source files in order to group related aspects.

I agree with you that the idea of grouping `Args` and `InputMethod` is a meaningful change since they are related.  But in contrast, violating against the alphabetical ordering just because those two data structures are related does not appear helpful to me.  At least in my opinion, this makes the overall navigation within the source file more complicated.

As an appropriate solution, I would like to suggest to resolve the arising issue of a harder navigation by applying at least one of the following strategies:

* Reflect the grouping in the documentation.  State in the file documentation how the data structures are ordered to outline your design decision.  The file documentation of the library crate root would thereby become something like a table of contents.
* Describe the relationship between the concerning data structures in their own documentation and outline the importance of them being situated next to each other in the source file.
* Outsource related data structures and logic to dedicated source files like the `sysexits` crate does.

File: src/lib.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (kevinmatthes) [reply src/lib.rs]: Okay. :+1:

File: src/lib.rs [hunk 8]
```diff
@@ -1,5 +1,262 @@
...
+
+    /// A set of strings to introduce some commits' messages.
+    message_starts_with: &'a [String],
+}
+
+impl Filter<'_> {
+    /// Whether the author's email address matches the expectations.
+    fn check_author_email(&self, email: &str) -> bool {
+        let contains =
+            self.email_contains.is_empty() || self.email_contains.iter().any(|e| email.contains(e));
```
Comment (kevinmatthes) [inline src/lib.rs:199]: I try to implement my `cargo-optimise` stricter regarding the Clippy lint levels, at the moment.  When running it over your code, it complained about obsolete closures.  Check what would happen with:

```rust
let contains = self.email_contains.is_empty() || self.email_contains.iter().any(email.contains);
```

**Review by kevinmatthes — CHANGES_REQUESTED** [review]
Thank you for your latest changes!  I left some feedback for you!

File: src/lib.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (wert007) [reply src/lib.rs]: What are the advantages in your opinion of alphabetical ordering in this context?

File: src/lib.rs [hunk 9]
```diff
@@ -1,5 +1,262 @@
...
+/// Parses the Git history.
+#[derive(Parser, Debug)]
+#[clap(version, about, long_about = None)]
+pub struct Args {
+    #[clap(subcommand)]
+    input_method: InputMethod,
+
+    /// Always show the entire output.
+    #[clap(short = 'v', long = "verbose")]
+    is_verbose: bool,
```
Comment (kevinmatthes) [inline src/lib.rs:16]: https://github.com/kevinmatthes/cargo-optimise uses different levels of verbosity.  Would this work for this case, as well?

File: src/lib.rs (hunk 8 above)
Comment (wert007) [reply src/lib.rs:199]:
```shell
error[E0615]: attempted to take value of method `contains` on type `&str`
   --> src\lib.rs:199:84
    |
199 |             self.email_contains.is_empty() || self.email_contains.iter().any(email.contains);
    |                                                                                    ^^^^^^^^ method, not a field
    |
help: use parentheses to call the method
    |
199 |             self.email_contains.is_empty() || self.email_contains.iter().any(email.contains(_));
    |                                                                                            +++

For more information about this error, try `rustc --explain E0615`.
```

does not compile

File: Cargo.lock [hunk 10]
```diff
@@ -21,24 +38,91 @@ dependencies = [
  "winapi",
 ]
 
+[[package]]
+name = "clap"
+version = "3.1.18"
```
Comment (kevinmatthes) [inline Cargo.lock:43]: Clap had major updates since you had been adding these dependencies.  When installing the application as described in its README, the compiler will yell at you since your code uses deprecated features, at the moment.

Run `cargo update` and fix arising issues like I did in https://github.com/kevinmatthes/cargo-optimise/commit/0f939ae6bc9d8d22a389936cbfd0cbddc4daca44:

* `bool`s need to be annotated `action`, now.
* If some options need to be parsed, they need to be annotated `value_parser`, now.

I recommend you to do this update since you will not only see the compiler warnings when installing but are also introducing a completely new dependency with this Pull Request.  So in my opinion, when merged, the Pull Request should only rely on state-of-the-art dependencies, this is, in our case, the latest available `clap`.

Comment (kevinmatthes) [reply Cargo.lock:43]: For further information on how to fix arising issues from the latest Clap, please consult the Clap tutorials in the official GitHub repository; this is also where I found the solution I implemented in the named commit.

File: src/lib.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (kevinmatthes) [reply src/lib.rs]: Its is easier to navigate through the code since you can rely on how the code is ordered.  Remember, not any editor offers you the possibility to jump to a certain scope to be chosen from a drop-down menu.

File: src/lib.rs (hunk 8 above)
Comment (kevinmatthes) [reply src/lib.rs:199]: I am sorry, I mixed up the functions.  The redundant closure is in line 104 before your changes, now line 361 after them:

```rust
.map(|l| l.loc())
```

Clippy also offers you some "help: replace the closure with the method itself: `LocDiff::loc`."  But you only get this with `#![deny(clippy::pedantic)]`.

File: src/lib.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (kevinmatthes) [reply src/lib.rs]: The other way around would be to separate the logic from `src/lib.rs` to source files on their applying the intended grouping.  You can then use the `src/lib.rs` for configuration.  Example:  https://github.com/kevinmatthes/cargo-optimise/blob/main/src/lib.rs

[the remark this answers is not in the document]
Comment (kevinmatthes) [reply src/lib.rs]: Since our `src/lib.rs` has grown quite big, I would like suggest to give the latter option a try.  Please consider to at least outsource the structs you grouped with your changes to a dedicated source file.  That way, their relationship becomes obvious and you can sort them by alphabet, again, in their source file.

File: src/lib.rs (hunk 9 above)
Comment (kevinmatthes) [reply src/lib.rs:16]: Okay, it was just a suggestion since the topics are related.

File: src/lib.rs (hunk 8 above)
Comment (wert007) [reply src/lib.rs:199]: I'm not sure, that this is really a better style, since you normally do not call member methods in that way `LocDiff::loc(l)`. I guess, that is why this lint is hidden behind `pedantic`.

File: Cargo.lock (hunk 10 above)
Comment (wert007) [reply Cargo.lock:43]: I updated to version `3.2.3` and it compiles and runs just fine. And since I took this code straight from their tutorial, I think it is still alright.

File: src/lib.rs [hunk 11]
```diff
@@ -1,5 +1,262 @@
...
+        self.is_verbose
+    }
+
+    /// Get the maximum duration between two commits considered spent working.
+    #[must_use]
+    pub fn duration(&self) -> u32 {
+        self.duration
+    }
+
+    /// Takes the output path, specified by the user, out of `Args`.
```
[the remark this answers is not in the document]
Comment (wert007) [reply src/lib.rs:82]: > * Imperative / Present instead of Indicative / Simple Present.  Otherwise you violate against the tense and mode of the other docstrings.

I do not quite get, where you see the simple present?

File: src/lib.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (kevinmatthes) [reply src/lib.rs]: > One: That's too bad for the person using the editor, maybe they should switch.

Now, I am very really interested:  tell me, which command line editor for Linux has scope selection like Visual Studio Code-like IDEs offer when a corresponding Language Server is installed, and supports at least the following languages ***WITH syntax highlighting***:

* Lox
* Haskell
* Fortran
* Asymptote
* GNU Octave
* GLSL
* Godot Script
* LaTeX

I admit that Lox is a very pathological example but I think you get the idea:  there are also people who do not use GUI-oriented editors.  Regarding Lox, I found an editor, https://github.com/curlpipe/ox, which allows adding syntax highlighting rules.  Also https://github.com/zee-editor/zee has options to add parsers, but Haskell is still in development and there is no Fortran at all.

Which terminal editor do you use?  Yours seems to have scope selection and searching with `Ctrl`+`F`.

> Second: Every editor has Ctrl+F, which easily allows finding members in a file.

`mcedit` does not:  the search function is triggered with `F7`.  Also `vi` / `vim` and `nano` do not support exactly this key binding but have search functionality at all.

> In general I'm not convinced that alphabetical ordering adds anything to a source file, [...]

Luck you.  I saw that much ***really*** bad source code in my life that I am glad that I can at least rely on something like how the code is ordered if it does not even compile without being rejected / complained about by multiple compilers, parsers, linters and Language Servers.  And the documentation system -- if it has a documentation to be created at all.  Such code is extremely hard to read and almost not maintainable.  If cannot even rely on the correctness of the code such that it produces the expected output, there is no chance editing it.

In summary:  any source file should have several points you can rely on.  Its sorting is one of them.

File: Cargo.lock (hunk 10 above)
Comment (kevinmatthes) [reply Cargo.lock:43]: If `cargo b` does not give you any warnings, everything will be indeed alright.

File: src/lib.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (kevinmatthes) [reply src/lib.rs]: > I do not know why you need this restriction.. :sweat_smile: You are also allowed to use multiple editors for different languages :smile:

Restrictions are always depending on the perspective.  I consider an editor which depends on a window server, this is the desktop environment of your system, a restriction since you cannot use it in SSH sessions or when the corresponding service is either down or does not exist at all.  Go to the first available TTY of your Linux system and try to edit a file.  If you should use Windows, imagine how a system would feel that only has a `cmd` or PowerShell as the sole possibility of interaction -- no mouse, only your keyboard can be used to enter commands.  If you just have a TTY and no window server available, you cannot start your text editor.  You should make sure to be able to work with one if you need to, for instance in case of an error in the configuration files of your desktop environment or if the GPU drivers are missing.

Using multiple editors for multiple languages just because of their syntax highlighting is not an option all.  Imagine `vi` / `vim` enthusiasts were forced to use Emacs just because of its syntax highlighting or vice versa.  Either an editor is capable of highlighting the syntax of all the languages you need or it is not an option to use at all.  The idea of an editor is not only to support you entering the text and highlighting the text, if possible, but also to be the one-and-only tool for editing texts.  If you come from Emacs, are you sure that you want to learn all the `vi` / `vim` command modes just for ***ONE*** language you cannot plugin into Emacs?  I suppose you will not do so.

> Adding alphabetical ordering would add nothing to a file, that does not compile. It would even take away the original intention of the author and make the code less readable.

Well, either you sort the structs or I am going to do so in my next Pull Request.  I do not see any reason to *not* sort the structs by alphabet.

[the remark this answers is not in the document]
Comment (wert007) [reply src/lib.rs]: > > I do not know why you need this restriction.. 😅 You are also allowed to use multiple editors for different languages 😄
> 
> Restrictions are always depending on the perspective. I consider an editor which depends on a window server, this is the desktop environment of your system, a restriction since you cannot use it in SSH sessions or when the corresponding service is either down or does not exist at all.

Then again, you could just use multiple editors depending on the occasion.. 😅 

> Go to the first available TTY of your Linux system and try to edit a file. If you should use Windows, imagine how a system would feel that only has a `cmd` or PowerShell as the sole possibility of interaction -- no mouse, only your keyboard can be used to enter commands. If you just have a TTY and no window server available, you cannot start your text editor. You should make sure to be able to work with one if you need to, for instance in case of an error in the configuration files of your desktop environment or if the GPU drivers are missing.

In most cases something similar to nano does completely suffice.

And if GPU drivers are missing, I recommend to first get those running, instead of programming in all those languages you listed 😄 

> Using multiple editors for multiple languages just because of their syntax highlighting is not an option all. Imagine `vi` / `vim` enthusiasts were forced to use Emacs just because of its syntax highlighting or vice versa. Either an editor is capable of highlighting the syntax of all the languages you need or it is not an option to use at all. The idea of an editor is not only to support you entering the text and highlighting the text, if possible, but also to be the one-and-only tool for editing texts. If you come from Emacs, are you sure that you want to learn all the `vi` / `vim` command modes just for _**ONE**_ language you cannot plugin into Emacs? I suppose you will not do so.

Well, if you have to be religious about your editor? It seems just to be a bad reason to narrow your horizon. In my experience everything, that claims to be a Swiss Army Knife without actually being one, is worse then the things actually designed to do the task at hand.
 
> > Adding alphabetical ordering would add nothing to a file, that does not compile. It would even take away the original intention of the author and make the code less readable.
> 
> Well, either you sort the structs or I am going to do so in my next Pull Request. I do not see any reason to _not_ sort the structs by alphabet.

Like I already explained, since `Args` and `InputMethod` are both interface to `clap` (and so is `Filter` in the way, that it is completely specified by the user), they are relevant to each other and should be close to each other. That is my reason, but I already explained it once. 😄

[the remark this answers is not in the document]
Comment (kevinmatthes) [reply src/lib.rs]: In order to return to the original purpose of this discussion:  let us just agree on that the sorting of symbols in a source file is a matter of taste to be discussed more in detail and / or applied in another Pull Request or issue.

**Comment by kevinmatthes** [comment]
By the way:  I took a look at the `slidesv3` project you referenced above.  It might be implemented even "rustier" and more portable with a `justfile` instead of all these Windows scripts.  The `justfile` is not considered a script at all but triggers the Makefile syntax highlighting of VS Code-like IDEs (as you seem to utilise as of your [`.gitignore`](https://github.com/wert007/slidesv3/blob/main/.gitignore)).

Further readings:

* https://github.com/sorairolake/sysexits-rs/blob/develop/justfile
* https://github.com/casey/just

File: src/lib.rs (hunk 11 above)
[the remark this answers is not in the document]
Comment (wert007) [reply src/lib.rs]: Actually decided to use the indicative/simple present everywhere, since that is what the rust docs seem to do.

**Comment by wert007** [comment]
> By the way: I took a look at the `slidesv3` project you referenced above. It might be implemented even "rustier" and more portable with a `justfile` instead of all these Windows scripts. The `justfile` is not considered a script at all but triggers the Makefile syntax highlighting of VS Code-like IDEs (as you seem to utilise as of your [`.gitignore`](https://github.com/wert007/slidesv3/blob/main/.gitignore)).
> 
> Further readings:
> 
> * https://github.com/sorairolake/sysexits-rs/blob/develop/justfile
> * https://github.com/casey/just

While this might be true, I have multiple reasons against that. For one this project is currently only developed on windows. It is also not really up to collaboration, so it does not really need to be portable at the time being. Also requiring `just` would add a not really necessary dependency, which I'm not really a fan of.

File: src/lib.rs [hunk 12]
```diff
@@ -1,5 +1,268 @@
...
 
+use std::path::PathBuf;
+
+use clap::{Parser, Subcommand};
+
+/// Parses the Git history.
+#[derive(Parser, Debug)]
+#[clap(version, about, long_about = None)]
+pub struct Args {
+    /// Specifies how the program will read the git history.
```
Comment (kevinmatthes) [inline src/lib.rs:11]: Git history

File: src/lib.rs [hunk 13]
```diff
@@ -1,5 +1,268 @@
...
+    #[clap(subcommand)]
+    input_method: InputMethod,
+
+    /// Always show the entire output.
+    #[clap(short = 'v', long = "verbose")]
+    is_verbose: bool,
+
+    /// Filter loc for certain file extension (e.g. `--file-extension cpp`). ORs if specified multiple times.
```
Comment (kevinmatthes) [inline src/lib.rs:19]: Filter the LOC diff for a certain file extension [...]

File: src/lib.rs [hunk 14]
```diff
@@ -1,5 +1,268 @@
...
+    #[clap(long)]
+    message_equals: Vec<String>,
+
+    /// Filter for certain commit messages. ORs if specified multiple times.
+    #[clap(short = 'l', long)]
+    message_starts_with: Vec<String>,
+}
+
+impl Args {
+    /// Gets the input method specified by the user.
```
Comment (kevinmatthes) [inline src/lib.rs:69]: Tense and mode not uniform:  above, there is still Imperative / Present.  "Filter [...]"

File: src/lib.rs [hunk 15]
```diff
@@ -1,5 +1,268 @@
...
+                    input.push_str(&buffer);
+                }
+
+                Ok(input)
+            }
+        }
+    }
+}
+
+/// The revealed filter creteria.
```
Comment (kevinmatthes) [inline src/lib.rs:164]: criteria

**Review by kevinmatthes — CHANGES_REQUESTED** [review]
There are still some small language issues.

File: Cargo.toml [hunk 16]
```diff
@@ -10,5 +10,5 @@ version = "0.1.0"
 
 [dependencies]
 chrono = "0.4.19"
-getopts = "0.2.21"
+clap = { version = "3.2.5", features = ["derive"] }
```
Comment (kevinmatthes) [inline Cargo.toml:13]: clap = { version = "3.2.6", features = [ "derive" ] }

Comment (kevinmatthes) [reply Cargo.toml:13]: We should rather think about setting up the depenabot, should we not?

File: src/lib.rs [hunk 17]
```diff
@@ -1,5 +1,268 @@
...
+    author_equals: &'a [String],
+
+    /// A set of substrings to be contained by some commits' hashes.
+    commit_contains: &'a [String],
+
+    /// A set of strings to match some commits' hashes.
+    commit_equals: &'a [String],
+
+    /// A set of substrings to be contained by some authors' email
+    /// addresses.
```
Comment (kevinmatthes) [inline src/lib.rs:182]: Why two lines?  Your docstrings above even longer and with the word "addresses" on the same line, you do not even exceed 80 characters!

File: src/lib.rs [hunk 18]
```diff
@@ -1,5 +1,268 @@
...
+    }
+
+    /// Gets the maximum duration between two commits considered spent working.
+    #[must_use]
+    pub fn duration(&self) -> u32 {
+        self.duration
+    }
+
+    /// Moves the output path specified by the user out of `Args` by calling
+    /// `Option::take`.
```
Comment (kevinmatthes) [inline src/lib.rs:88]: "... out of `Args`."  Move the way this is achieved to the detailed description below.

File: src/lib.rs [hunk 19]
```diff
@@ -471,7 +530,7 @@ impl LocDiff {
         &self.file
     }
 
-    /// Calculate the LOC diff.
+    /// Calculates the LOC diff.
```
Comment (kevinmatthes) [inline src/lib.rs:533]: Revert this change.  It does not add anything to the source file but raises an unnecessary grammar issue:  there is no subject, now that you use Indicative instead of Imperative!

File: src/main.rs [hunk 20]
```diff
@@ -1,135 +1,19 @@
...
-        None => 3,
-        Some(Ok(it)) => it,
-        Some(Err(error)) => {
-            eprintln!("Invalid duration: {error}!");
-            return sysexits::ExitCode::Usage;
-        }
-    };
-    let path = matches.opt_str("output");
-    let commits = match input.read() {
+    let mut args = commit_analyzer::Args::parse();
```
Comment (kevinmatthes) [inline src/main.rs:6]: Just for completeness:  how you made the application show the help immediately when no subcommand was passed?  Is there any Clap trick to do so implicitly?

**Review by kevinmatthes — CHANGES_REQUESTED** [review]
Here are still some open points I would to ask you to change.

File: src/main.rs (hunk 20 above)
Comment (wert007) [reply src/main.rs:6]: I just copied the examples. Maybe this works, because this application has subcommands, but I do not know.. 😅

File: src/lib.rs (hunk 19 above)
Comment (wert007) [reply src/lib.rs:533]: grammarly does not seem to have a problem with this.

File: Cargo.toml (hunk 16 above)
Comment (wert007) [reply Cargo.toml:13]: I do not think we need to update clap that frequently. (Or probably at all tbh) 😅

File: src/lib.rs (hunk 19 above)
Comment (kevinmatthes) [reply src/lib.rs:533]: In my opinion, this is a phrase without a subject.  You just have a verb and an object.  In Imperative, this would not matter since this mode does not require a subject in each case.

Comment (wert007) [reply src/lib.rs:533]: Well, I checked it with a grammar checker. So I don't know what to tell you.. 😅
